### DISCUSSION TITLE
surround all requests by try-catch bloc

### DISCUSSION THREAD

**Discussion by shabakett** [canton7/RestEase#236] [Q&A]
how to auto surround all requests by try-catch bloc?
example : force
```
var result1 = await api1.Method1Async(param1, param2);
```
to execute as:
```
try
{
	return await api1.Method1Async(param1, param2);
}
catch (Exception x)
{
	Log.AppendLine(x.Message);
	if (x is HttpRequestException c && c.StatusCode == null) return "Check your Internet connection";
	return x.Message;
}
```

**Comment by canton7** [comment]
One fairly common pattern is something like:

```cs
private async Task<string> DoRequest(Func<ISomeApi, Task<string>> method)
{
	try
	{
		return await method(api1);
	}
	catch (Exception x)
	{
		Log.AppendLine(x.Message);
		if (x is HttpRequestException c && c.StatusCode == null) return "Check your Internet connection";
		return x.Message;
	}
}
```

Then:

```cs
var result1 = await DoRequest(x => x.Method1Async(param1, param2));
```

Another option is to use a [RequestModifier](https://github.com/canton7/RestEase#requestmodifier), and catch exceptions in there. The downside is that you still need to return a `HttpResponseMessage`, rather than a string.

**Reply by shabakett** [reply]
please post an example on how to use the RequestModifier with HttpClient in this case

**Reply by canton7** [reply]
Oops, no, a `RequestModifier won't help here. A [Custom HttpClient](https://github.com/canton7/RestEase#custom-httpclient) might though.

```cs
public class CustomHttpClientHandler : WebRequestHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        try
        {
            return await base.SendAsync(request, cancellationToken)o
        }
        catch (Exception e)
        {
            return <something suitable>;
        }
    }
}
```

**Reply by shabakett** [reply]
I think this is the most appropriate way, just I use `public class CustomHttpClientHandler : HttpClientHandler` and in the catch block I return `new HttpResponseMessage { Content = new StringContent(e.Message) }`
